File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    # DP -- precompute left, right max
    def trap(self, height):
        if len(height) <= 2:
            return 0
        total = 0
        left, right = [], []
        leftMax, rightMax = 0, 0
        for i in range(len(height)):
            leftMax = max(height[i], leftMax)
            left.append(leftMax)
        for i in range(len(height) - 1, -1, -1):
            rightMax = max(height[i], rightMax)
            right.append(rightMax)
        right.reverse()
        for i in range(1, len(height) - 1):
            area = min(left[i - 1], right[i + 1])
            total += 0 if area - height[i] < 0 else area - height[i]
        
        return total
    # # brute force - min of max of left and right at each element
    # def trap(self, height: List[int]) -> int:
    #     if len(height) <= 2:
    #         return 0
    #     totalTrapped = 0
    #     for i in range(1, len(height) - 1):
    #         h = min(max(height[:i]), max(height[i+1:]))
    #         area = h - height[i]
    #         totalTrapped += 0 if area < 0 else area
    #     return totalTrapped

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        
        if(n < 3)
            return 0;
        
        vector<int> left(n), right(n);
        
        int currMax = 0;
        for(int i = 0 ; i < n; i++){
            left[i] = currMax;
            currMax = max(height[i], currMax);
        }
        
        currMax = 0;
        for(int i = n - 1; i >= 0; i--){
            right[i] = currMax;
            currMax = max(height[i], currMax);
        }
        
        int trapped = 0;
        for(int i = 1; i < n - 1; i++){
            trapped += max(min(left[i], right[i]) - height[i], 0);
        }
        
        return trapped;
    }
};